File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/rest/CommonRESTHandlers.java
package com.wearezeta.auto.common.rest;

import org.apache.commons.lang3.StringUtils;
import java.util.logging.Logger;

import com.wearezeta.auto.common.log.ZetaLogger;

import javax.ws.rs.ProcessingException;
import javax.ws.rs.client.Entity;
import javax.ws.rs.client.Invocation.Builder;
import javax.ws.rs.core.GenericType;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.function.Supplier;

public final class CommonRESTHandlers {

    private RESTResponseHandler responseHandler;
    private int maxRetries = 1;

    public CommonRESTHandlers(RESTResponseHandler responseHandler) {
        this.responseHandler = responseHandler;
    }

    public CommonRESTHandlers(RESTResponseHandler responseHandler, int maxRetries) {
        this.responseHandler = responseHandler;
        this.maxRetries = maxRetries;
    }

    private static final Logger log = ZetaLogger.getLog(CommonRESTHandlers.class.getSimpleName());
    private static final String EMPTY_LOG_RECORD = "EMPTY";
    private static final int IS_ALIVE_VERIFICATION_TIMEOUT_MS = 5000;

    public static boolean isAlive(URL siteURL) {
        try {
            final HttpURLConnection connection = (HttpURLConnection) siteURL.openConnection();
            connection.setRequestMethod("HEAD");
            connection.setConnectTimeout(IS_ALIVE_VERIFICATION_TIMEOUT_MS);
            connection.setReadTimeout(IS_ALIVE_VERIFICATION_TIMEOUT_MS);
            connection.connect();
            final int responseCode = connection.getResponseCode();
            log.info(String.format("Response code from %s: %s", siteURL.toString(), responseCode));
            return (responseCode == HttpURLConnection.HTTP_OK);
        } catch (Exception e) {
            // Just ignore
            // e.printStackTrace();
        }
        return false;
    }

    private static final int MAX_SINGLE_ENTITY_LENGTH_IN_LOG = 400;

    private static String formatLogRecord(Object entity) {
        String result = EMPTY_LOG_RECORD;
        if (entity != null) {
            if (entity instanceof String) {
                result = ((String) entity);
            } else {
                result = entity.toString();
            }
            result = result.isEmpty() ? EMPTY_LOG_RECORD :
                    StringUtils.abbreviate(result, MAX_SINGLE_ENTITY_LENGTH_IN_LOG);
        }
        return result;
    }

    private Response retryRequest(Supplier<Response> requestFunc) {
        int tryNum = 0;
        ProcessingException savedException;
        do {
            try {
                return requestFunc.get();
            } catch (ProcessingException e) {
                savedException = e;
            }
        } while (++tryNum < this.maxRetries);
        throw savedException;
    }

    public <T> T httpPost(Builder webResource, Object entity,
                          Class<T> responseEntityType, int[] acceptableResponseCodes) {
        return httpPost(webResource, MediaType.APPLICATION_JSON, entity,
                responseEntityType, acceptableResponseCodes);
    }

    public <T> T httpPost(Builder webResource, String contentType,
                          Object entity, Class<T> responseEntityType, int[] acceptableResponseCodes) {
        log.info("POST REQUEST...");
        log.info(String.format(" >>> Input data: %s", formatLogRecord(entity)));
        final Response response = retryRequest(
                () -> webResource.post(Entity.entity(entity, contentType), Response.class)
        );
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes);
    }

    public Response httpPostFullResponse(Builder webResource, String contentType,
                                         Object entity) throws RESTError {
        log.info("POST REQUEST...");
        log.info(String.format(" >>> Input data: %s", formatLogRecord(entity)));
        final Response response = retryRequest(
                () -> webResource.post(Entity.entity(entity, contentType), Response.class)
        );
        return response;
    }

    public <T> T getEntityFromResponse(Response response, Class<T> responseEntityType, int[] acceptableResponseCodes)
            throws RESTError {
        T responseEntity;
        try {
            response.bufferEntity();
            responseEntity = response.readEntity(responseEntityType);
            log.info(String.format(" >>> Response: %s", formatLogRecord(responseEntity)));
            this.responseHandler.verifyRequestResult(response.getStatus(),
                    acceptableResponseCodes, formatLogRecord(responseEntity));
        } catch (ProcessingException | IllegalStateException | NullPointerException e) {
            responseEntity = null;
            log.warning(e.getMessage());

            if (!"java.lang.String".equals(responseEntityType.getName())) {
                try {
                    String responseString = response.readEntity(String.class);
                    log.info(String.format(" >>> Response: %s", formatLogRecord(responseString)));
                    this.responseHandler.verifyRequestResult(
                            response.getStatus(), acceptableResponseCodes, formatLogRecord(responseString));
                } catch (ProcessingException | IllegalStateException | NullPointerException ex) {
                    log.warning(ex.getMessage());
                }
            }
        }
        return responseEntity;
    }

    public String httpPost(Builder webResource, Object entity, int[] acceptableResponseCodes) {
        String returnString = httpPost(webResource, entity, String.class, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public String httpPost(Builder webResource, Object entity, String contentType, int[] acceptableResponseCodes) {
        String returnString = httpPost(webResource, contentType, entity, String.class, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public <T> T httpPut(Builder webResource, Object entity,
                         Class<T> responseEntityType, int[] acceptableResponseCodes) {
        log.info("PUT REQUEST...");
        log.info(String.format(" >>> Input data: %s", formatLogRecord(entity)));
        final Response response = retryRequest(
                () -> webResource.put(
                        Entity.entity(entity, MediaType.APPLICATION_JSON_TYPE),
                        Response.class)
        );
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes);
    }

    public String httpPut(Builder webResource, Object entity, int[] acceptableResponseCodes) {
        String returnString = httpPut(webResource, entity, String.class, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public <T> T httpDelete(Builder webResource, Object entity, Class<T> responseEntityType,
                            int[] acceptableResponseCodes) {
        log.info("DELETE REQUEST...");
        final Response response = retryRequest(
                () -> webResource.method("DELETE", Entity.entity(entity, MediaType.APPLICATION_JSON_TYPE),
                        Response.class)
        );
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes);
    }

    public String httpDelete(Builder webResource, Object entity, int[] acceptableResponseCodes) {
        String returnString = httpDelete(webResource, entity, String.class, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public String httpDelete(Builder webResource, int[] acceptableResponseCodes) {
        String returnString = httpDelete(webResource, null, String.class, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public <T> T httpGet(Builder webResource,
                         GenericType<T> responseEntityType, int[] acceptableResponseCodes) {
        log.info("GET REQUEST...");
        final Response response = retryRequest(
                () -> webResource.get(Response.class)
        );
        T responseEntity;
        try {
            response.bufferEntity();
            responseEntity = response.readEntity(responseEntityType);
            log.info(String.format(" >>> Response: %s", formatLogRecord(responseEntity)));
            this.responseHandler.verifyRequestResult(response.getStatus(),
                    acceptableResponseCodes, formatLogRecord(responseEntity));
        } catch (ProcessingException | IllegalStateException | NullPointerException e) {
            responseEntity = null;
            log.warning(e.getMessage());
            if (!"java.lang.String".equals(responseEntityType.getRawType().getName())) {
                try {
                    String responseString = response.readEntity(String.class);
                    log.info(String.format(" >>> Response: %s", formatLogRecord(responseString)));
                    this.responseHandler.verifyRequestResult(
                            response.getStatus(), acceptableResponseCodes, formatLogRecord(responseString));
                } catch (ProcessingException | IllegalStateException | NullPointerException ex) {
                    log.warning(ex.getMessage());
                }
            }
        }
        return responseEntity;
    }

    public String httpGet(Builder webResource, int[] acceptableResponseCodes) {
        String returnString = httpGet(webResource, new GenericType<String>() {}, acceptableResponseCodes);
        returnString = returnString == null ? "" : returnString;
        return returnString;
    }

    public int getResponseCode(Builder webResource, String contentType, Object entity) {
        log.info("POST REQUEST...");
        log.info(String.format(" >>> Input data: %s", formatLogRecord(entity)));
        return retryRequest(
                () -> webResource.post(Entity.entity(entity, contentType), Response.class)
        ).getStatus();
    }

    public int getResponseCode(Builder webResource) {
        log.info("GET REQUEST...");
        return retryRequest(
                () -> webResource.get(Response.class)
        ).getStatus();
    }
}

File: tests/testsCore/src/main/kotlin/com/wire/android/tests/core/utils/CommonRESTHandlers.kt
/*
 * Wire
 * Copyright (C) 2025 Wire Swiss GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see http://www.gnu.org/licenses/.
 */
package com.wire.android.tests.core.utils

import com.wire.android.tests.core.exceptions.RESTError
import com.wire.android.tests.core.services.RESTResponseHandler
import org.apache.commons.lang3.StringUtils
import java.net.HttpURLConnection
import java.net.URL
import java.util.function.Supplier
import java.util.logging.Logger
import javax.ws.rs.ProcessingException
import javax.ws.rs.client.Entity
import javax.ws.rs.client.Invocation.Builder
import javax.ws.rs.core.GenericType
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

class CommonRESTHandlers @JvmOverloads constructor(
    private val responseHandler: RESTResponseHandler,
    private val maxRetries: Int = 1
) {
    companion object {
        private val log: Logger = ZetaLogger.getLog(CommonRESTHandlers::class.simpleName)
        private const val EMPTY_LOG_RECORD = "EMPTY"
        private const val IS_ALIVE_VERIFICATION_TIMEOUT_MS = 5000
        private const val MAX_SINGLE_ENTITY_LENGTH_IN_LOG = 400

        fun isAlive(siteURL: URL): Boolean {
            return try {
                (siteURL.openConnection() as HttpURLConnection).apply {
                    requestMethod = "HEAD"
                    connectTimeout = IS_ALIVE_VERIFICATION_TIMEOUT_MS
                    readTimeout = IS_ALIVE_VERIFICATION_TIMEOUT_MS
                    connect()
                }.let { connection ->
                    val responseCode = connection.responseCode
                    log.info("Response code from ${siteURL}: $responseCode")
                    responseCode == HttpURLConnection.HTTP_OK
                }
            } catch (e: Exception) {
                false
            }
        }

        private fun formatLogRecord(entity: Any?): String {
            if (entity == null) return EMPTY_LOG_RECORD

            val result = when (entity) {
                is String -> entity
                else -> entity.toString()
            }

            return if (result.isEmpty()) EMPTY_LOG_RECORD
            else StringUtils.abbreviate(result, MAX_SINGLE_ENTITY_LENGTH_IN_LOG)
        }
    }

    private fun retryRequest(requestFunc: Supplier<Response>): Response {
        var tryNum = 0
        var savedException: ProcessingException? = null
        do {
            try {
                return requestFunc.get()
            } catch (e: ProcessingException) {
                savedException = e
            }
        } while (++tryNum < maxRetries)
        throw savedException!!
    }

    fun <T> httpPost(
        webResource: Builder,
        entity: Any?,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ): T = httpPost(webResource, MediaType.APPLICATION_JSON, entity, responseEntityType, acceptableResponseCodes)

    fun <T> httpPost(
        webResource: Builder,
        contentType: String,
        entity: Any?,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ): T {
        log.info("POST REQUEST...")
        log.info(" >>> Input data: ${formatLogRecord(entity)}")
        val response = retryRequest {
            webResource.post(Entity.entity(entity, contentType), Response::class.java)
        }
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes)
    }

    fun httpPostFullResponse(
        webResource: Builder,
        contentType: String,
        entity: Any?
    ): Response {
        log.info("POST REQUEST...")
        log.info(" >>> Input data: ${formatLogRecord(entity)}")
        return retryRequest {
            webResource.post(Entity.entity(entity, contentType), Response::class.java)
        }
    }

    fun <T> getEntityFromResponse(
        response: Response,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ): T {
        return try {
            response.bufferEntity()
            val responseEntity = response.readEntity(responseEntityType)
            log.info(" >>> Response: ${formatLogRecord(responseEntity)}")
            responseHandler.verifyRequestResult(
                response.status,
                acceptableResponseCodes,
                formatLogRecord(responseEntity)
            )
            responseEntity
        } catch (e: ProcessingException) {
            handleResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Failed to process response", 500)
        } catch (e: IllegalStateException) {
            handleResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Invalid response state", 400)
        } catch (e: NullPointerException) {
            handleResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Null response", 500)
        }
    }

    private fun <T> handleResponseError(
        response: Response,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ) {
        if (responseEntityType.name != "java.lang.String") {
            try {
                val responseString = response.readEntity(String::class.java)
                log.info(" >>> Response: ${formatLogRecord(responseString)}")
                responseHandler.verifyRequestResult(
                    response.status,
                    acceptableResponseCodes,
                    formatLogRecord(responseString))
            } catch (ex: Exception) {
                log.warning(ex.message)
            }
        }
    }

    fun httpPost(
        webResource: Builder,
        entity: Any?,
        acceptableResponseCodes: IntArray
    ): String = httpPost(webResource, entity, String::class.java, acceptableResponseCodes) ?: ""

    fun httpPost(
        webResource: Builder,
        entity: Any?,
        contentType: String,
        acceptableResponseCodes: IntArray
    ): String = httpPost(webResource, contentType, entity, String::class.java, acceptableResponseCodes) ?: ""

    fun <T> httpPut(
        webResource: Builder,
        entity: Any?,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ): T {
        log.info("PUT REQUEST...")
        log.info(" >>> Input data: ${formatLogRecord(entity)}")
        val response = retryRequest {
            webResource.put(Entity.entity(entity, MediaType.APPLICATION_JSON), Response::class.java)
        }
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes)
    }

    fun httpPut(
        webResource: Builder,
        entity: Any?,
        acceptableResponseCodes: IntArray
    ): String = httpPut(webResource, entity, String::class.java, acceptableResponseCodes) ?: ""

    fun <T> httpDelete(
        webResource: Builder,
        entity: Any?,
        responseEntityType: Class<T>,
        acceptableResponseCodes: IntArray
    ): T {
        log.info("DELETE REQUEST...")
        val response = retryRequest {
            webResource.method("DELETE", Entity.entity(entity, MediaType.APPLICATION_JSON), Response::class.java)
        }
        return getEntityFromResponse(response, responseEntityType, acceptableResponseCodes)
    }

    fun httpDelete(
        webResource: Builder,
        entity: Any?,
        acceptableResponseCodes: IntArray
    ): String = httpDelete(webResource, entity, String::class.java, acceptableResponseCodes) ?: ""

    fun httpDelete(
        webResource: Builder,
        acceptableResponseCodes: IntArray
    ): String = httpDelete(webResource, null, String::class.java, acceptableResponseCodes) ?: ""

    fun <T> httpGet(
        webResource: Builder,
        responseEntityType: GenericType<T>,
        acceptableResponseCodes: IntArray
    ): T? {
        log.info("GET REQUEST...")
        val response = retryRequest {
            webResource.get(Response::class.java)
        }
        return try {
            response.bufferEntity()
            val responseEntity = response.readEntity(responseEntityType)
            log.info(" >>> Response: ${formatLogRecord(responseEntity)}")
            responseHandler.verifyRequestResult(
                response.status,
                acceptableResponseCodes,
                formatLogRecord(responseEntity))
                        responseEntity
        } catch (e: ProcessingException) {
            handleGetResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Failed to process response", 500)
        } catch (e: IllegalStateException) {
            handleGetResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Invalid response state", 400)
        } catch (e: NullPointerException) {
            handleGetResponseError(response, responseEntityType, acceptableResponseCodes)
            throw RESTError("Null response", 500)
        }
    }

    private fun <T> handleGetResponseError(
        response: Response,
        responseEntityType: GenericType<T>,
        acceptableResponseCodes: IntArray
    ) {
        if (responseEntityType.type == String::class.java){
            try {
                val responseString = response.readEntity(String::class.java)
                log.info(" >>> Response: ${formatLogRecord(responseString)}")
                responseHandler.verifyRequestResult(
                    response.status,
                    acceptableResponseCodes,
                    formatLogRecord(responseString))
            } catch (ex: Exception) {
                log.warning(ex.message)
            }
        }
    }

    fun httpGet(
        webResource: Builder,
        acceptableResponseCodes: IntArray
    ): String = httpGet(webResource, GenericType<String>(String.Companion::class.java), acceptableResponseCodes) ?: ""

    fun getResponseCode(
        webResource: Builder,
        contentType: String,
        entity: Any?
    ): Int {
        log.info("POST REQUEST...")
        log.info(" >>> Input data: ${formatLogRecord(entity)}")
        return retryRequest {
            webResource.post(Entity.entity(entity, contentType), Response::class.java)
        }.status
    }

    fun getResponseCode(webResource: Builder): Int {
        log.info("GET REQUEST...")
        return retryRequest {
            webResource.get(Response::class.java)
        }.status
    }
}
